mesh/mesh.cpp

```diff
@@ -11277,6 +11277,66 @@ GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
    }
 }
 
+GeometricFactors::GeometricFactors(const GridFunction *nodes_,
+                                   const IntegrationRule &ir,
+                                   int flags, MemoryType d_mt)
+{
+   this->mesh = nullptr;
+   IntRule = &ir;
+   computed_factors = flags;
+
+   const GridFunction *nodes = nodes_;
+   const FiniteElementSpace *fespace = nodes->FESpace();
+   const FiniteElement *fe = fespace->GetFE(0);
+   const int dim  = fe->GetDim();
+   const int vdim = fespace->GetVDim();
+   const int NE   = fespace->GetNE();
+   const int ND   = fe->GetDof();
+   const int NQ   = ir.GetNPoints();
+
+   unsigned eval_flags = 0;
+   MemoryType my_d_mt = (d_mt != MemoryType::DEFAULT) ? d_mt :
+                        Device::GetDeviceMemoryType();
+   if (flags & GeometricFactors::COORDINATES)
+   {
+      X.SetSize(vdim*NQ*NE, my_d_mt); // NQ x SDIM x NE
+      eval_flags |= QuadratureInterpolator::VALUES;
+   }
+   if (flags & GeometricFactors::JACOBIANS)
+   {
+      J.SetSize(dim*vdim*NQ*NE, my_d_mt); // NQ x SDIM x DIM x NE
+      eval_flags |= QuadratureInterpolator::DERIVATIVES;
+   }
+   if (flags & GeometricFactors::DETERMINANTS)
+   {
+      detJ.SetSize(NQ*NE, my_d_mt); // NQ x NE
+      eval_flags |= QuadratureInterpolator::DETERMINANTS;
+   }
+
+   const QuadratureInterpolator *qi = fespace->GetQuadratureInterpolator(ir);
+   // All X, J, and detJ use this layout:
+   qi->SetOutputLayout(QVectorLayout::byNODES);
+   // Same as UsesTensorBasis(*fespace) but we already have GetFE(0):
+   const bool use_tensor_products = dynamic_cast<const TensorBasisElement*>(fe)
+                                    != nullptr;
+   qi->DisableTensorProducts(!use_tensor_products);
```

Comment:
```suggestion
   qi->DisableTensorProducts(!UsesTensorBasis(*fespace));
```

Reply:
I propagated this change to `GeometricFactors(const Mesh *mesh, const IntegrationRule &ir, int flags, MemoryType d_mt) ` as well. For consistency.